### PR TITLE
Api patch [semver:patch]

### PR DESCRIPTION
**Pull request by eddiewebb**
### Checklist

- [ ] All new jobs, commands, executors, parameters have descriptions
- [ ] Examples have been added for any significant new features
- [ ] README has been updated, if necessary

### Description

Whereas PIPELINE NUMBER is the only consistent, incremental value that spans jobs invoked by a single webhook event (containing a collection of commits).  it is mapped to `buildNumber` which is repeated builds of the same project in jira model

Whereas REPO_NAME is the only consistent value that spans all activities of a project, it maps to "pipeline ID", which is like a build project in jira model

Where as RETRIES should be the same pipelined, but replace build, we get this for free since pipeline remains the same [1] [2] through workfklows that are retries.

Where as `updateSequenceNumber` is some incremental indicator of freshness, we use epoch.

Where as, all this fits 💯  with the UI of our product.

## Builds
Here lots of builds, jira shows latest

#### Circleci
Screen Shot 2020-02-28 at 8 33 51 PM

#### Jira
Screen Shot 2020-02-28 at 8 33 59 PM

#### Circleci

Screen Shot 2020-02-28 at 8 42 55 PM

#### JIra
Screen Shot 2020-02-28 at 8 42 39 PM

[1](https://circleci.com/api/v2/workflow/d1688623-aa5b-4427-b640-b9d58a6a96cd)
[2](https://circleci.com/api/v2/workflow/27dd93d1-c3d2-42ff-8041-134cadf05f44)

### CODE REVIEW

File: src/commands/notify.yml [hunk 1]
```diff
@@ -52,17 +57,26 @@ steps:
         fi
 
         run () {
+          verify_token
           parse_jira_key_array
-          if [ "x" != "x${ISSUE_KEYS}" ]; then
+          HAS_JSD_SERVICE_ID = << parameters.service_id >>
```
Comment (eddiewebb) [inline src/commands/notify.yml]: gotta fix this, its nit failing build but blowing uyp/errring out

File: src/commands/notify.yml [hunk 2]
```diff
@@ -103,7 +117,8 @@ steps:
           echo {} | jq \
           --arg time_str "$(date +%s)" \
           --arg lastUpdated "${iso_time}" \
-          --arg pipeline "${CIRCLE_PROJECT_REPONAME}" \
+          --arg pipelineNumber "${CIRCLE_PIPELINE_NUMBER:-\<< pipeline.number >>}" \
```
Comment (eddiewebb) [inline src/commands/notify.yml]: this is a trick, we escape the parameter so orb packing doesnt repalce it but build processing will.  Its wrapped by a variable to enable unit testing.

File: src/commands/notify.yml [hunk 3]
```diff
@@ -173,14 +190,23 @@ steps:
...
+                "updateSequenceNumber": $time_str,
                 "displayName": $deployDisplay,
                 "description": $description,
                 "url": $url,
                 "state": $state,
                 "lastUpdated": $lastUpdated,
-                "issueKeys": $issueKeys,
+                "associations": [
+                  {
+                    "associationType": "issueKeys",
```
Comment (eddiewebb) [inline src/commands/notify.yml]: i dont know where you got issueIdorKeys, but irt must have been a hint, not the actual value.  We're passing keys here, and I assumed ID below, but you should follow up wioth atlassian.

Comment (eddiewebb) [reply src/commands/notify.yml]:
```{
  "acceptedDeployments": [
    {
      "pipelineId": "circleci-samples",
      "environmentId": "64983647689364-non-prod",
      "deploymentSequenceNumber": 1
    }
  ],
  "rejectedDeployments": [],
  "unknownIssueKeys": [],
  "unknownAssociations": [
    {
      "associationType": "serviceIdOrKeys",
      "values": [
        "testid"
      ]
    }
  ]
}
```

Comment (KyleTryon) [reply src/commands/notify.yml]: This comes from the new undocumented API. Sending a slack DM.

**Comment by eddiewebb** [comment]
Any job failure (https://circleci.com/workflow-run/da85474a-36e1-4240-a525-d8e808a29112) marks jira failure https://eddiewebb.atlassian.net/browse/CC-26

passing workflows show too. You can see these values are incremented. You can re-run either workflow and the jira should show thagt the number stayed the same, but status will update.


I think we should merge this alone, but I also just realized that we can use the workflows `job` endpoint and possible work out dependencies to send `pending` instead of success for intermediate jobs. (its sugar since Jira will only reflect during length of workflow)

File: src/commands/notify.yml [hunk 4]
```diff
@@ -79,22 +107,18 @@ steps:
         }
 
         check_workflow_status () {
+          URL="https://circleci.com/api/v2/workflow/${CIRCLE_WORKFLOW_ID}"
+          fetch $URL /tmp/workflow.json
+          export WORKFLOW_STATUS=$(jq -r '.status' /tmp/workflow.json)
+          export CIRCLE_PIPELINE_NUMBER=$(jq -r '.pipeline_number' /tmp/workflow.json)
```
Comment (eddiewebb) [inline src/commands/notify.yml]: @KyleTryon - we can pull pipeline number when we check workflow status.

File: src/commands/notify.yml [hunk 5]
```diff
@@ -124,9 +149,9 @@ steps:
             "builds": [
               {
                 "schemaVersion": "1.0",
-                "pipelineId": $pipeline,
-                "buildNumber": 1,
-                "updateSequenceNumber": $buildNumber,
+                "pipelineId": $projectName,
+                "buildNumber": $pipelineNumber,
+                "updateSequenceNumber": $time_str,
```
Comment (eddiewebb) [inline src/commands/notify.yml]: just use epoch

**Comment by eddiewebb** [comment]
Screen Shot 2020-02-28 at 8 33 59 PM
Screen Shot 2020-02-28 at 8 33 51 PM

**Comment by eddiewebb** [comment]
I marked this as major, but then dropped all the way to patch, here's why.

1. Build numbers will jump from 1 to (something greater than 1) so data from this orb is guaranteed to work with existing data. 
2. it is important people adopt this.
3. I think Jira changed and the old data wont show up anymore.


The one risk to move back to major:

the UI description they see in jira will still say `#xxx your project` but `#xxx` will jump from latets build_num to latest pipeline_num, which is almost guarenteed to be _lower_ than build.  That may be confusing, but since it should squash all other, it should be ok.

**Comment by eddiewebb** [comment]
@KyleTryon  - I can collab tomorrow (Tuesday) if you want, anytime before 10 or after 1

**Comment by KyleTryon** [comment]
> @KyleTryon - I can collab tomorrow (Tuesday) if you want, anytime before 10 or after 1

Heck ya 🎉 . I believe it is very close but happy to collab (especially given the time crunch). I am available at the same times as well

File: tests/cases/simple_with_service_ID.yml [hunk 6]
```diff
@@ -0,0 +1,10 @@
+jobs:
+  build:
+    jobs:
+    docker:
+    - image: circleci/node:10
+    steps:
+      - run: echo "hello"
+      - jira/notify:
+          state_path:  /tmp/jira.status
+          service_id: "b:YXJpOmNsb3VkOmdyYXBoOjpzZXJ2aWNlL2E1MDU4MjMyLTZmNzctMTFlYS1hOGRmLTBhNzdmM2Y0NTMwNC84NjIwNmQwNC03MmNhLTExZWEtOTU2OC0wYTc3ZjNmNDUzMDQ="
```
Comment (eddiewebb) [inline tests/cases/simple_with_service_ID.yml]: the Change Managemetn (service IDs) is only on deployments. Not buids.

File: tests/test.bats [hunk 7]
```diff
@@ -156,3 +143,105 @@ function setup {
...
+  # then is passes
+  [[ "$status" == "0" ]]
+
+  # and reports success 
+  assert_jq_match '.deployments | length' 1 /tmp/jira-status.json
+  assert_jq_match '.deployments[0].deploymentSequenceNumber' 313 /tmp/jira-status.json
+  assert_jq_match '.deployments[0].pipeline.id' "${CIRCLE_PROJECT_REPONAME}" /tmp/jira-status.json
+}
+
+@test "8: Basic expansion with serviceID" {
```
Comment (eddiewebb) [inline tests/test.bats]: We can knock this off, the new format is not supported with service ID, and not used by builds at all, so this test is really replicating one above.

**Comment by eddiewebb** [comment]
@KyleTryon I realize I cant just slack you anymore, lol . I am free until 2 and again at 3:30.  Send an invite to ewebb     @     atlassian.com

**Comment by KyleTryon** [comment]
@eddiewebb my tests are failing locally but I think that has something to do with my environment. I pushed up the changes we discussed and included a test specifically to check for an invalid response when an invalid service ID is sent. 

Would love if you could give this a final look. I would also like to confirm 100% that the output in JSD we are seeing is in fact what we want to see.

Free for the rest of the day as well if you are available. 

Here is the latest passing test: https://app.circleci.com/pipelines/github/CircleCI-Public/jira-connect-orb/174/workflows/34271a6f-9e6b-4d41-bd0b-c6335be20798/jobs/384/artifacts

**Pull request merged**